### PR TITLE
EHN: multi-column explode

### LINKED ISSUE #39240: ENH: Multi-Column explode
#### Is your feature request related to a problem?

I have often looked for the option to use the `explode` function on multiple columns at the same time. 

Example:
```
 A | B |    C   |    D
------------------------
 1 | a | [1,2] | [7,8]
 2 | b | [3,4] | [9,1]
 3 | c | [5,6] | [2,3]
```
===>
```
 A | B |    C  |   D
------------------
 1 | a |    1   |   7
 1 | a |    2   |  8
 2 | b |    3   |  9
 2 | b |    4   |   1
 3 | c |    5   |   2
 3 | c |    6   |   3
```
For obvious reasons exploding first one and then the other column will not yield the result we want here, because we will end up duplicating the values in the first explode during the second explode.

#### Describe the solution you'd like

Either it could be `df.explode(["C","D"])` or it could be its own function `df.multi_explode(["C","D"])`. Both ways could use more or less the same method that the existing explode function uses.

#### API breaking implications

The only issue I see with this are cases where the lists in the C and D column aren't the same length, in which case we would need to pad them.

#### Context
I don't know if this has been discussed previously, but if there is interest in this a Pull request concerning this feature I could try to implement it.

### PR DESCRIPTION
**Pull request by iynehz**
- [x] closes #39240
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -7969,32 +7971,61 @@ def explode(
...
+        3    4  1    e
+        """
         if not self.columns.is_unique:
             raise ValueError("columns must be unique")
+        if is_scalar(column) or isinstance(column, tuple):
+            columns = [column]
+        elif isinstance(column, list) and all(
+            map(lambda c: is_scalar(c) or isinstance(c, tuple), column)
+        ):
+            if len(column) > len(set(column)):
```
Comment (phofl) [inline pandas/core/frame.py:8011]: Please document this

File: pandas/core/frame.py [hunk 2]
```diff
@@ -7969,32 +7971,61 @@ def explode(
...
+            # "List[Union[str, Tuple[Any, ...], List[Union[str, Tuple[Any, ...]]]]]")
+            columns = column  # type: ignore[assignment]
+        else:
+            raise ValueError("column must be a scalar, tuple, or list thereof")
+
+        mylen = lambda x: len(x) if is_list_like(x) else -1
+        counts0 = self[columns[0]].apply(mylen)
+        for c in columns[1:]:
+            if not all(counts0 == self[c].apply(mylen)):
+                raise ValueError("columns must have matching element counts")
```
Comment (phofl) [inline pandas/core/frame.py:8024]: Document too

File: pandas/core/frame.py [hunk 3]
```diff
@@ -7969,32 +7971,61 @@ def explode(
...
+            if len(column) > len(set(column)):
+                raise ValueError("column must be unique")
+            # mypy: Incompatible types in assignment (expression has type
+            # "List[Union[str, Tuple[Any, ...]]]", variable has type
+            # "List[Union[str, Tuple[Any, ...], List[Union[str, Tuple[Any, ...]]]]]")
+            columns = column  # type: ignore[assignment]
+        else:
+            raise ValueError("column must be a scalar, tuple, or list thereof")
+
+        mylen = lambda x: len(x) if is_list_like(x) else -1
```
Comment (phofl) [inline pandas/core/frame.py:8020]: This is probably a performance killer, so please do this only if len(columns) > 1. Maybe there is a better way to do this? Also add asv for this case

File: pandas/core/frame.py [hunk 4]
```diff
@@ -7969,32 +7971,61 @@ def explode(
...
+        counts0 = self[columns[0]].apply(mylen)
+        for c in columns[1:]:
+            if not all(counts0 == self[c].apply(mylen)):
+                raise ValueError("columns must have matching element counts")
 
         df = self.reset_index(drop=True)
-        result = df[column].explode()
-        result = df.drop([column], axis=1).join(result)
+        result = DataFrame({c: df[c].explode() for c in columns})
```
Comment (phofl) [inline pandas/core/frame.py:8027]: This adds overhead for len = 1 case, let's look at asvs here

File: pandas/tests/frame/methods/test_explode.py [hunk 5]
```diff
@@ -9,13 +9,23 @@ def test_error():
...
         df.explode(list("AA"))
 
     df.columns = list("AA")
     with pytest.raises(ValueError, match="columns must be unique"):
         df.explode("A")
 
+    df1 = df.assign(C=[["a", "b", "c"], "foo", [], ["d", "e", "f"]])
+    df1.columns = list("ABC")
+    with pytest.raises(ValueError, match="columns must have matching element counts"):
+        df1.explode(list("AC"))
```
Comment (phofl) [inline pandas/tests/frame/methods/test_explode.py:27]: Add test where one side is scalar and other side is list

File: pandas/tests/frame/methods/test_explode.py [hunk 6]
```diff
@@ -180,3 +190,24 @@ def test_explode_sets():
     result = df.explode(column="a").sort_values(by="a")
     expected = pd.DataFrame({"a": ["x", "y"], "b": [1, 1]}, index=[1, 1])
     tm.assert_frame_equal(result, expected)
+
+
+def test_multi_columns():
+    df = pd.DataFrame(
```
Comment (phofl) [inline pandas/tests/frame/methods/test_explode.py:196]: add gh reference

File: pandas/core/frame.py [hunk 7]
```diff
@@ -7969,32 +7971,61 @@ def explode(
...
+            columns = [column]
+        elif isinstance(column, list) and all(
+            map(lambda c: is_scalar(c) or isinstance(c, tuple), column)
+        ):
+            if len(column) > len(set(column)):
+                raise ValueError("column must be unique")
+            # mypy: Incompatible types in assignment (expression has type
+            # "List[Union[str, Tuple[Any, ...]]]", variable has type
+            # "List[Union[str, Tuple[Any, ...], List[Union[str, Tuple[Any, ...]]]]]")
+            columns = column  # type: ignore[assignment]
```
Comment (phofl) [inline pandas/core/frame.py:8016]: Could you type columns to avoid mypy error?

**Review by phofl — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py [hunk 8]
```diff
@@ -7953,32 +7963,67 @@ def explode(self, column: str | tuple, ignore_index: bool = False) -> DataFrame:
...
+        3    3  1    d
+        3    4  1    e
+        """
         if not self.columns.is_unique:
             raise ValueError("columns must be unique")
 
+        columns: list[str | tuple]
+        if is_scalar(column) or isinstance(column, tuple):
+            # mypy: List item 0 has incompatible type "Union[str, Tuple[Any, ...],
+            # List[Union[str, Tuple[Any, ...]]]]"; expected
```
Comment (jreback) [inline pandas/core/frame.py:8002]: you might be able to cast to remove this warning

File: pandas/core/frame.py [hunk 9]
```diff
@@ -7953,32 +7963,67 @@ def explode(self, column: str | tuple, ignore_index: bool = False) -> DataFrame:
...
+            # "Union[str, Tuple[Any, ...]]"
+            columns = [column]  # type: ignore[list-item]
+        elif isinstance(column, list) and all(
+            map(lambda c: is_scalar(c) or isinstance(c, tuple), column)
```
Comment (jreback) [inline pandas/core/frame.py:8006]: why is this 2nd part needed? is it tested?

File: pandas/core/frame.py [hunk 10]
```diff
@@ -7953,32 +7963,67 @@ def explode(self, column: str | tuple, ignore_index: bool = False) -> DataFrame:
...
+        ):
+            if len(column) == 0:
```
Comment (jreback) [inline pandas/core/frame.py:8008]: `if not len(column)`

File: pandas/core/frame.py [hunk 11]
```diff
@@ -7910,16 +7910,23 @@ def stack(self, level: Level = -1, dropna: bool = True):
...
         """
         Transform each element of a list-like to a row, replicating index values.
 
         .. versionadded:: 0.25.0
 
         Parameters
         ----------
-        column : str or tuple
-            Column to explode.
+        column : str or tuple or list thereof
```
Comment (jreback) [inline pandas/core/frame.py:7925]: we should consider deprecating & renaming to subset to match others, can you open an issue

File: pandas/core/frame.py [hunk 12]
```diff
@@ -7953,32 +7963,67 @@ def explode(self, column: str | tuple, ignore_index: bool = False) -> DataFrame:
...
+        elif isinstance(column, list) and all(
+            map(lambda c: is_scalar(c) or isinstance(c, tuple), column)
+        ):
+            if len(column) == 0:
+                raise ValueError("column must be nonempty")
+            if len(column) > len(set(column)):
+                raise ValueError("column must be unique")
+            columns = column
+        else:
+            raise ValueError("column must be a scalar, tuple, or list thereof")
```
Comment (jreback) [inline pandas/core/frame.py:8014]: tests hit this?

Comment (jreback) [inline pandas/core/frame.py:8014]: why are we accepting tuples? for a column MI?

File: pandas/core/frame.py [hunk 13]
```diff
@@ -7953,32 +7963,67 @@ def explode(self, column: str | tuple, ignore_index: bool = False) -> DataFrame:
...
+
         df = self.reset_index(drop=True)
-        result = df[column].explode()
-        result = df.drop([column], axis=1).join(result)
+        if len(columns) == 1:
+            result = df[column].explode()
+        else:
+            mylen = lambda x: len(x) if is_list_like(x) else -1
+            counts0 = self[columns[0]].apply(mylen)
```
Comment (jreback) [inline pandas/core/frame.py:8021]: just apply to everything, then you can iterate, this is pretty awkward

File: pandas/tests/frame/methods/test_explode.py [hunk 14]
```diff
@@ -9,13 +9,34 @@ def test_error():
     df = pd.DataFrame(
         {"A": pd.Series([[0, 1, 2], np.nan, [], (3, 4)], index=list("abcd")), "B": 1}
     )
-    with pytest.raises(ValueError, match="column must be a scalar"):
+    with pytest.raises(
```
Comment (jreback) [inline pandas/tests/frame/methods/test_explode.py:12]: this test is getting big, can you split into multiple ones (ok to rename the original), parameterize is good as well if posilbe

File: pandas/tests/frame/methods/test_explode.py [hunk 15]
```diff
@@ -180,3 +201,25 @@ def test_explode_sets():
...
     tm.assert_frame_equal(result, expected)
+
+
+def test_multi_columns():
+    # GH 39240
+    df = pd.DataFrame(
+        {
+            "A": pd.Series([[0, 1, 2], np.nan, [], (3, 4)], index=list("abcd")),
+            "B": 1,
+            "C": [["a", "b", "c"], "foo", [], ["d", "e"]],
```
Comment (jreback) [inline pandas/tests/frame/methods/test_explode.py:212]: what about nan elements in C as well?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 10 above)
Comment (phofl) [reply pandas/core/frame.py:8008]: if not colum should be enough?

File: pandas/core/frame.py (hunk 9 above)
Comment (stphnlyd) [reply pandas/core/frame.py:8006]: to make sure every element in the list has the right type. I will add a test.

File: pandas/core/frame.py (hunk 12 above)
Comment (stphnlyd) [reply pandas/core/frame.py:8014]: yes that's for MI. Actually here it's just that the ValueError message before change does not mention tuple, but the code logic already supports str or tuple.

File: pandas/core/frame.py (hunk 13 above)
Comment (stphnlyd) [reply pandas/core/frame.py:8021]: make sense

File: pandas/core/frame.py (hunk 10 above)
Comment (stphnlyd) [reply pandas/core/frame.py:8008]: I see PEP8 recommends `if not seq`. That should have minimal number of operations on the bytecode level.

File: pandas/core/frame.py (hunk 12 above)
Comment (stphnlyd) [reply pandas/core/frame.py:8014]: yes I think I have it

File: pandas/core/frame.py (hunk 8 above)
Comment (phofl) [reply pandas/core/frame.py:8010]: I think you can use assert here, this would remove the mypy warning

**Comment by phofl** [comment]
small comments, otherwise lgtm

**Comment by stphnlyd** [comment]
@phofl I've pushed to address your comments

**Comment by jreback** [comment]
@stphnlyd pls merge master

File: pandas/tests/frame/methods/test_explode.py [hunk 16]
```diff
@@ -9,14 +9,61 @@ def test_error():
...
+                "B": 1,
+                "C": [["a", "b", "c"], "foo", [], "d"],
+            },
+            list("abcd"),
+            list("AC"),
+            "columns must have matching element counts",
+        ),
+    ],
+)
+def test_error_multi_columns(input_dict, input_index, input_subset, error_message):
```
Comment (phofl) [inline pandas/tests/frame/methods/test_explode.py:60]: Can you parametrize only the changing parts?

File: pandas/tests/frame/methods/test_explode.py [hunk 17]
```diff
@@ -180,3 +227,65 @@ def test_explode_sets():
...
+                    ["d", "e"],
+                    ["d", "e"],
+                    np.nan,
+                ],
+            },
+            list("aaabcdde"),
+        ),
+    ],
+)
+def test_multi_columns(
```
Comment (phofl) [inline pandas/tests/frame/methods/test_explode.py:284]: Similar here

File: pandas/tests/frame/methods/test_explode.py (hunk 16 above)
Comment (stphnlyd) [reply pandas/tests/frame/methods/test_explode.py:60]: This test_error_multi_columns() is a new function I added in my PR, not something that already exists there. I added one more test case to this test_error_multi_columns() in this commit.

File: pandas/tests/frame/methods/test_explode.py (hunk 17 above)
Comment (stphnlyd) [reply pandas/tests/frame/methods/test_explode.py:284]: similar as explained above

File: pandas/tests/frame/methods/test_explode.py (hunk 16 above)
Comment (phofl) [reply pandas/tests/frame/methods/test_explode.py:60]: That was not what I was referring to. Most of your parametrization values are the same for all cases, so no need to put them into the parametrization

Comment (stphnlyd) [reply pandas/tests/frame/methods/test_explode.py]: @phofl updated. btw "Python Dev / actions-310-dev" fails coverage upload, but that's not caused by me..

**Comment by stphnlyd** [comment]
@phofl any comments?

File: pandas/core/frame.py [hunk 18]
```diff
@@ -8168,16 +8168,27 @@ def stack(self, level: Level = -1, dropna: bool = True):
...
         ----------
-        column : str or tuple
-            Column to explode.
+        column : str or tuple or list thereof
+            Column(s) to explode.
+            For multiple columns, specify a non-empty list with each element
+            be str or tuple, and all specified columns their list-like data
+            on same row of the frame must have matching length.
+
+            .. versionadded:: 1.3.0
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:8189]:
```suggestion
            .. versionadded:: 1.4.0
```

**Comment by simonjayhawkins** [comment]
@stphnlyd can you add a release note. doc/source/whatsnew/v1.4.0.rst

**Comment by stphnlyd** [comment]
@simonjayhawkins thanks I've squashed the PR and now target 1.4.0, and I've added release note. The PR checks failed but that's not caused by my changes.

**Review by jreback — CHANGES_REQUESTED** [review]
looks good. is the case where an element is nan in one of the multi-columns and a list in the other one tested / handled?

**Comment by jreback** [comment]
@phofl good here?

**Comment by phofl** [comment]
Yes lgtm

File: doc/source/whatsnew/v1.4.0.rst [hunk 19]
```diff
@@ -29,7 +29,7 @@ enhancement2
 
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
--
+- :meth:`DataFrame.explode` now supports exploding multiple columns. Its ``column`` argument now also accepts a list of str or tuples for exploding on multiple columns at the same time (:issue:`39240`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.4.0.rst:32]: ok putting this in 1.3, if you can move and ping on green.

**Review by jreback — APPROVED** [review]

**Comment by stphnlyd** [comment]
@jreback targeting 1.3 now

**Comment by jreback** [comment]
thanks @stphnlyd very nice

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Pull request merged**